### PR TITLE
Multi-agent Intersection

### PR DESCRIPTION
**Pull request by lqy0057**
## What changes do you make in this PR?

Multi-agent unprotected intersection is now available

## Checklist

* [x] I have merged the latest main branch into current branch.
* [x] I have run `bash scripts/format.sh` before merging.
* Please use "squash and merge" mode.

### CODE REVIEW

File: pgdrive/envs/marl_envs/marl_intersection.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+            new_map = MAIntersectionMap(self.pg_world, map_config)
+            self.maps[self.current_seed] = new_map
+            self.current_map = self.maps[self.current_seed]
+
+    def _update_destination_for(self, vehicle):
+        # when agent re-joined to the game, call this to set the new route to destination
+        end_roads = copy.deepcopy(self.spawn_roads)
+        end_roads.remove(vehicle.current_road)
+        end_road = -get_np_random(self._DEBUG_RANDOM_SEED).choice(end_roads)  # Use negative road!
+        vehicle.routing_localization.set_route(vehicle.lane_index[0], end_road.end_node)
```
Comment (pengzhenghao) [inline pgdrive/envs/marl_envs/marl_intersection.py:85]: Why this function is different from MARound?

File: pgdrive/envs/marl_envs/marl_intersection.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        if all(d.values()):
+            env.reset()
+        if (s + 1) % 100 == 0:
+            print(
+                "Finish {}/10000 simulation steps. Time elapse: {:.4f}. Average FPS: {:.4f}".format(
+                    s + 1,
+                    time.time() - start, (s + 1) / (time.time() - start)
+                )
+            )
+    print(f"(PGDriveEnvV2) Total Time Elapse: {time.time() - start}")
```
Comment (pengzhenghao) [inline pgdrive/envs/marl_envs/marl_intersection.py:272]:
```suggestion
    print(f"(MAIntersection) Total Time Elapse: {time.time() - start}")
```

File: pgdrive/envs/marl_envs/marl_intersection.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    finally:
+        env.close()
+
+
+if __name__ == "__main__":
+    # _draw()
+    _vis()
+    # _vis_debug_respawn()
+    # _profiwdle()
+    # _long_run()
```
Comment (pengzhenghao) [inline pgdrive/envs/marl_envs/marl_intersection.py:336]: So you have tried all these functions and they all work well? (If so, then close this comment.)

File: pgdrive/scene_creator/vehicle_module/routing_localization.py [hunk 4]
```diff
@@ -81,7 +81,15 @@ def update(self, map: Map, start_road_node=None, final_road_node=None, random_se
...
-            socket = get_np_random(random_seed).choice(map.blocks[-1].get_socket_list())
+            sockets = map.blocks[-1].get_socket_list()
+            while True:
+                socket = get_np_random(random_seed).choice(sockets)
+                if not socket.is_socket_node(start_road_node):
+                    break
+                else:
+                    sockets.remove(socket)
+                    if len(sockets) == 0:
+                        raise ValueError("Can not set a destination!")
```
Comment (pengzhenghao) [inline pgdrive/scene_creator/vehicle_module/routing_localization.py:92]: In which case this error would happen? Can we make sure that case will never happens? (If so then close this comment)

File: pgdrive/tests/test_env/test_ma_intersection.py [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+    if not (set(obs.keys()) == set(reward.keys()) == set(env.observation_space.spaces.keys())):
+        raise ValueError
+    assert env.observation_space.contains(obs)
+    assert isinstance(reward, dict)
+    assert isinstance(info, dict)
+    assert isinstance(done, dict)
+    return obs, reward, done, info
+
+
+def test_ma_roundabout_env():
```
Comment (pengzhenghao) [inline pgdrive/tests/test_env/test_ma_intersection.py]:
```suggestion
def test_ma_intersection_env():
```

File: pgdrive/tests/test_env/test_ma_intersection.py [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+            for step in range(100):
+                act = {k: [1, 1] for k in env.vehicles.keys()}
+                o, r, d, i = _act(env, act)
+                if step == 0:
+                    assert not any(d.values())
+        finally:
+            env.close()
+
+
+def test_ma_roundabout_horizon():
```
Comment (pengzhenghao) [inline pgdrive/tests/test_env/test_ma_intersection.py]:
```suggestion
def test_ma_intersection_horizon():
```

File: pgdrive/tests/test_env/test_ma_intersection.py [hunk 7]
```diff
@@ -0,0 +1,707 @@
...
+                        assert r[kkk] == -777
+
+                if d["__all__"]:
+                    break
+                last_keys = new_keys
+        finally:
+            env.close()
+
+
+def test_ma_roundabout_reset():
```
Comment (pengzhenghao) [inline pgdrive/tests/test_env/test_ma_intersection.py]:
```suggestion
def test_ma_intersection_reset():
```

File: pgdrive/tests/test_env/test_ma_intersection.py [hunk 8]
```diff
@@ -0,0 +1,707 @@
...
+                    print("Finish {} agents. Success {} agents.".format(agent_count, success_count))
+                    o = env.reset()
+                    assert env.observation_space.contains(o)
+                    _check_spaces_after_reset(env, o)
+                    break
+    finally:
+        env.close()
+
+
+def test_ma_roundabout_close_spawn():
```
Comment (pengzhenghao) [inline pgdrive/tests/test_env/test_ma_intersection.py]:
```suggestion
def test_ma_intersection_close_spawn():
```

**Comment by pengzhenghao** [comment]
image


is this class deprecated?

**Comment by pengzhenghao** [comment]
A special bug happens, please run marl_intersection.py right now!

**Comment by pengzhenghao** [comment]
image

There are one vehicle alive, but I have make sure it is a vehicle that crash others and keep half body in sky, so we can't accurate it at that time!

**Comment by pengzhenghao** [comment]
We should add some criterion to terminate those agent who (1) has very high Z position, or .. I don't know if there are better way to determine an un-controllable agent.

File: pgdrive/envs/marl_envs/marl_intersection.py (hunk 1 above)
Comment (lqy0057) [reply pgdrive/envs/marl_envs/marl_intersection.py:85]: No circular route to drive to reverse road in the intersection block

Comment (lqy0057) [reply pgdrive/envs/marl_envs/marl_intersection.py:85]: The modification in set_route function is also used to solve this problem.

**Comment by lqy0057** [comment]
> image
> 
> is this class deprecated?

Yes DeprecatedWarning please

File: pgdrive/envs/marl_envs/marl_intersection.py (hunk 1 above)
Comment (pengzhenghao) [reply pgdrive/envs/marl_envs/marl_intersection.py:85]: Why not~~

Comment (pengzhenghao) [reply pgdrive/envs/marl_envs/marl_intersection.py:85]: We should allow U-turn too!

**Comment by pengzhenghao** [comment]
> > image
> > is this class deprecated?
> 
> Yes DeprecatedWarning please

Solved!

**Comment by pengzhenghao** [comment]
Merge when tests passed!

**Pull request merged**